### PR TITLE
This is PR 450 plus further changes to fix a out-of-memory conditions

### PR DESCRIPTION
**Pull request by sickpig**

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -5483,22 +5538,25 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
...
+        //   Validate that INVs are a valid type and not null.
+        if (vInv.size() > MAX_INV_SZ || vInv.empty())
         {
+            LOCK(cs_main);
             Misbehaving(pfrom->GetId(), 20);
             return error("message inv size() = %u", vInv.size());
         }
-        
-        for (unsigned int nInv = 0; nInv < vInv.size(); nInv++)  // Validate that INVs are a valid type
+        for (unsigned int nInv = 0; nInv < vInv.size(); nInv++)
```
Comment (greatwolf) [inline src/main.cpp:5551]: shouldn't `size_t` be preferred here?

File: src/main.cpp [hunk 2]
```diff
@@ -5483,22 +5538,25 @@ bool ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, in
...
         {
             const CInv &inv = vInv[nInv];
-            if (!((inv.type == MSG_TX) || (inv.type == MSG_BLOCK)))
+            if (!((inv.type == MSG_TX) || (inv.type == MSG_BLOCK)) || inv.hash.IsNull())
             {
```
Comment (greatwolf) [inline src/main.cpp:5555]: Can reduce the nested parens by changing this to `inv.type != MSG_TX && inv.type != MSG_BLOCK` instead.

**Comment by sickpig** [comment]
@greatwolf thanks for the review.

**Comment by sickpig** [comment]
Superseded by PR #468

**Pull request closed without merging**